src/zenml/zen_stores/migrations/README.md

```diff
@@ -0,0 +1,51 @@
+# How to create a migration
+
+## How alembic is works
+
+src/zenml/alembic.ini is used for the configuration of alembic in general. The
+env.py file is the entrypoint for alembic where we are able to influence some
+settings, like for example the location of our sqlite db.
+
+Now when you create an alembic revision using `alembic revision ...` the 
+database instance will be compared to the state of all SQLModels under 
+src/zenml/zen_stores/schemas. This means, in order for alembic to work properly
+for you, make sure you have a fully instantiated version of the `develop` state
```

Comment:
Instead of doing multiple revisions for each PR/change, does it make sense to instead do a single revision per release? Or would the speedup of this be negligible?

Reply:
I had a rather long debate with Stefan about this. Here is the reason that convinced me

Let's say Stefan has a PR with a Migration "3ak2l", on top of that you have a PR with a Migration "4owl42". Whenever we push these into develop, our dev servers are updated through these migrations and the version is stored. If we now combine these two migrations into a release migration "v0.20.5" we lose the history ... ->  "3ak2l" -> "4owl42" -> ...
and replace it with ... -> "v0.20.5" -> ... - This will lead to problems if we now try to upgrade our dev servers again, our dev server will be at version "4owl42" and alembic will not know anything about this version. So we will have to nuke our dev servers after every release. This same effect will also impact all of us during development.

Reply:
So which migrations to apply is determined by the migration tag in alembic, not by the state of the DB schema?

Reply:
We defined a release workflow that will give us the best of both worlds and resolve any branches: https://www.notion.so/zenml/Make-A-Release-adb5e415107e4189bbb894a4c08c00f1#8027d6d8331243f49cf24e3c0b9abaed

This will now create an empty revision at release time that gives us a version stamp matching our code version.

Reply:
After more debating with Alexej, we found that alembic has a nice feature called "branches" that allow developers to add migration scripts in parallel, then merge them at a later point, if necessary. We made the alembic merge a mandatory step in the release process. This also allows us to "stamp" the state of the migration scripts with every release, so we can run things like "alembic downgrade 0.20.5" and "alembic upgrade 0.20.6" manually, if necessary.

Reply:
That sounds great. I also think that database migration will be rather a sporadic event in the future and something we consider carefully instead of everyone simply modifying them on random branches ;)